CHANGES.md

```diff
@@ -58,7 +59,34 @@ NTRU Prime sntrup761 and X25519 with SHA-512: sntrup761x25519-sha512](https://ww
 
 ## Behavioral changes and enhancements
 
-* [GH-468](https://github.com/apache/mina-sshd/issues/468) SFTP: validate length of data received: must not be more than requested
+### [GH-461](https://github.com/apache/mina-sshd/issues/461) Fix heartbeats with `wantReply=true`
+
+The client-side heartbeat mechanism has been changed. Such heartbeats are configured with an interval
+(`CoreModuleProperties.HEARTBEAT_INTERVAL`) > 0. Previously they could also be configured with a timeout
+(`CoreModuleProperties.HEARTBEAT_REPLY_WAIT`). If the timeout was <= 0, the client would just send a
+heartbeat request, but not expect any answer. However, with a timeout > 0, it would send the request
+with a flag telling the server to send back a reply, and then wait for the given duration for the
+reply to arrive, and terminate the connection if no reply was forthcoming.
+
+That could cause trouble if the timeout was fairly long, and the server was slow to respond. If the
+timeout was longer than the interval it could also delay subsequent heartbeats.
+
+Newly, `CoreModuleProperties.HEARTBEAT_REPLY_WAIT` is _deprecated_. There is a new configuration property
+`CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX` instead. It defines a limit for how many heartbeats without
+reply a session survives. If the value is <= 0, the client still sends heartbeats without expecting
+any answer. If > 0, it will request a reply from the server for each heartbeat message, and it will
+terminate the connection once there are `CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX` previous heartbeats
+for which no reply was received yet.
+
+This new way to configure heartbeats corresponds to the configuration in OpenSSH via the
+`ServerAliveInterval` and `ServerAliveCountMax` options.
+
+For compatibility with old configurations that do define `CoreModuleProperties.HEARTBEAT_REPLY_WAIT` > 0,
+the new code maps this to the new configuration (but only if `CoreModuleProperties.HEARTBEAT_INTERVAL` > 0
+and the new property `CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX` has _not_ been set) by setting
+`CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX = (CoreModuleProperties.HEARTBEAT_REPLY_WAIT / CoreModuleProperties.HEARTBEAT_INTERVAL) + 1`.
```

Comment:
With AI help, I have the following reworded doc

The client-side heartbeat mechanism has been updated.

Heartbeats are primarily configured using the CoreModuleProperties.HEARTBEAT_INTERVAL interval. If this value is > 0, heartbeats will be sent to the server.

Previously, heartbeats could also be configured with the CoreModuleProperties.HEARTBEAT_REPLY_WAIT timeout. If the timeout was <= 0, the client would send a heartbeat request without expecting a reply. If the timeout was > 0, the client would send the request with a flag indicating that the server should reply. The client would then wait for the specified duration for the reply and terminate the connection if no reply was received.

This old mechanism could cause issues if the timeout was long and the server was slow to respond. A timeout longer than the interval could also delay subsequent heartbeats.

The CoreModuleProperties.HEARTBEAT_REPLY_WAIT property is now deprecated.

It has been replaced by a new configuration property, CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX. This property defines the limit for the number of heartbeats sent without receiving a reply before a session is terminated. If the value is <= 0, the client sends heartbeats without expecting any reply. If the value is > 0, the client requests a reply from the server for each heartbeat message and terminates the connection if the number of unanswered heartbeats reaches CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX.

This new configuration method aligns with the OpenSSH configuration options ServerAliveInterval and ServerAliveCountMax.

For compatibility with older configurations where CoreModuleProperties.HEARTBEAT_REPLY_WAIT > 0, the new code maps this to the new configuration (only if CoreModuleProperties.HEARTBEAT_INTERVAL > 0 and CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX is not set) by setting CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX to (CoreModuleProperties.HEARTBEAT_REPLY_WAIT / CoreModuleProperties.HEARTBEAT_INTERVAL) + 1.

Reply:
Done. I've taken this as inspiration and have rephrased the description along these lines.